cmd/lakectl/cmd/retry_client.go

```diff
@@ -0,0 +1,71 @@
+package cmd
+
+import (
+	"context"
+	"crypto/x509"
+	"net/http"
+	"net/url"
+	"regexp"
+	"time"
+
+	"github.com/hashicorp/go-retryablehttp"
+)
+
+var (
+	// these errors aren't typed, so we match by regexp
+	redirectsErrorRe  = regexp.MustCompile(`stopped after \d+ redirects\z`)
+	schemeErrorRe     = regexp.MustCompile(`unsupported protocol scheme`)
+	notTrustedErrorRe = regexp.MustCompile(`certificate is not trusted`)
+)
+
+func NewRetryClient(maxRetries int, retryWaitMin, retryWaitMax time.Duration, transport *http.Transport) *http.Client {
+	retryClient := retryablehttp.NewClient()
+	if transport != nil {
+		retryClient.HTTPClient.Transport = transport
+	}
+	retryClient.Logger = nil
+	retryClient.RetryMax = maxRetries
+	retryClient.RetryWaitMin = retryWaitMin
+	retryClient.RetryWaitMax = retryWaitMax
+	retryClient.CheckRetry = lakectlRetryPolicy
+	return retryClient.StandardClient()
+}
+
+// lakectl retry policy - we retry in the following cases:
+// HTTP status 429 - too many requests
+// HTTP status 500 - internal server error - could be recoverable
+// HTTP status 503 - service unavailable
+// We retry on all client transport errors except for:
+//   - too many redirects
+//   - invalid http scheme/protocol
+//   - TLS cert verification failure
+func lakectlRetryPolicy(ctx context.Context, resp *http.Response, err error) (bool, error) {
+	// do not retry on context.Canceled or context.DeadlineExceeded
+	if ctx.Err() != nil {
+		return false, ctx.Err()
+	}
+
+	// handle client transport errors
+	if err != nil {
+		if v, ok := err.(*url.Error); ok {
```

Comment:
If it's not `*url.Error` then we'll always retry? Can you explain why in a comment?

Reply:
Added a comment to explain the logic.